tests/fixtures/mapping-destinations.yml

```diff
@@ -28,14 +30,40 @@ tools:
         - pulsar
       require:
         - inherited
+  three_core_test_tool:
+    scheduling:
+      require:
+        - local
+    rules:
+      - if: input_size > 3
+        id: arbitrary_3_core_assignment
+        cores: 3
 
 destinations:
   local:
     cores: 4
     mem: 16
     scheduling:
+      accept:
+        - local
       prefer:
         - general
+    rules:
+      - if: |
+          from tpv.core.entities import Tool
+          is_three_core_job = False
+          try:
+            [tool_entity] = list(filter(lambda x: isinstance(x, Tool), entities))
+            is_three_core_job = tool_entity.cores == 3
```

Comment:
This code reveals some interesting design issues, but I'm not sure there are clean way to solve them.

Firstly, `tool_entity.cores` works here because it's not an expression. But if it had been an expression, then this would not really have worked.

Secondly, it would have been ideal to just have been able to simplify this entire block to `if: entity.cores == 3`, but we can't, because `if` conditions are evaluated twice, once in evaluate_early and again in evaluate_late. In evaluate_early, we don't yet have the combined entity, and therefore, `entity.cores` still refers to the destination entity only.

To make it work, we would have to have two types of `if` conditions. `pre_combine_if` and `post_combine_if`. That would be very ugly, hence why the `if` is evaluated twice instead. We can't avoid the `evaluate_early` step either, since that what makes core and mem comparisons prior to combining possible. 

One possible solution, is to say that core, mem and gpu clauses do not support expressions. In that case, we can completely eliminate the evaluate_early step and make this work. Thoughts?

Reply:
At the time of making a list of candidate destinations for ranking, the cores, men and GPUs for the job are known. Could the combined entity at this point be inserted into the context?

Reply:
I tried a major redesign, but it didn't go well. The main issue is that, since rules are able to modify core and mem count, they need to be evaluated prior to combining, and therefore, there's no clean way to do it in a delayed fashion. It ends up complicating things enormously, even if it's achievable. I ended up doing it the way you suggested. Let me know what you think.